### PR TITLE
optimize store.DeleteCollection

### PR DESCRIPTION
**Pull request by microyahoo**
Signed-off-by: Liang Zheng <zhengliang0901@gmail.com>

As discussion at https://github.com/kubernetes/kubernetes/pull/105606#issuecomment-949458100 , the code can be simplified to achieve the same effect.

#### What type of PR is this?
/kind cleanup

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?
NONE

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by dims** [comment]
@fedebongio looks like this still needs to be looked at...

**Comment by caesarxuchao** [comment]
cc @sxllwx

**Comment by sxllwx** [comment]
@caesarxuchao  Nice catch! Thanks for looking into this :) 

> I guess this won't change the peak cpu/memory usage, but will make the total execution time of the function longer.

I agree. Goroutines are frequently created and destroyed during execution, which is less efficient than consuming from chan.

But the current code is simpler and easier to understand,so  I agree with this PR.

@microyahoo can you provide some benchmark results for reference?

File: staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go [hunk 1]
```diff
@@ -1124,59 +1124,43 @@ func (e *Store) DeleteCollection(ctx context.Context, deleteValidation rest.Vali
 	if workersNumber < 1 {
 		workersNumber = 1
 	}
+
```
Comment (wojtek-t) [inline staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go:1127]: This change looks correct to me, but I have one concern.

This effectively is creating a goroutine for every item in a list (which potentially may mean thousands or even more of goroutines). Starting too many goroutines quickly has caused us issues in the past, so if we decide to do that, I would like to see that:
- this actually improves something (some benchmark or sth)
- some explanation that starting too man goroutines at once is not a problem (golang folks were planning to work on that in the past so it's probably it's no longer really a problem at that scale)

**Pull request closed without merging**